File: cpp/a11/net/wire_stream.cc
// Copyright 2026 The A11 Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "a11/net/wire_stream.h"

#include <cstddef>
#include <string>
#include <utility>

#include <absl/status/status.h>
#include <absl/status/status_macros.h>
#include <absl/strings/ascii.h>

#include "a11/data/types.h"

namespace a11::net {

absl::Status WireStreamOptions::Validate() const {
  if (max_buffered_incoming_messages < 1 ||
      max_buffered_incoming_messages > 1024) {
    return absl::InvalidArgumentError(
        "max_buffered_incoming_messages must be in [1, 1024]");
  }
  const size_t minimum = data::EmptyWireMessageSize();
  if (max_single_message_size < minimum ||
      max_single_message_size > kMaxSingleMessageSize) {
    return absl::InvalidArgumentError(
        "max_single_message_size is outside the supported range");
  }
  if (max_buffered_incoming_bytes < minimum) {
    return absl::InvalidArgumentError(
        "max_buffered_incoming_bytes is smaller than an empty message");
  }
  if (message_timeout < absl::ZeroDuration() &&
      message_timeout != absl::InfiniteDuration()) {
    return absl::InvalidArgumentError(
        "message_timeout must be non-negative or infinite");
  }
  return absl::OkStatus();
}

absl::StatusOr<data::ByteMap> NormalizeWireHeaders(data::ByteMap headers) {
  data::ByteMap normalized;
  for (auto& [key, value] : headers) {
    std::string folded = absl::AsciiStrToLower(key);
    ABSL_RETURN_IF_ERROR(data::ValidateName(folded));
    normalized.insert_or_assign(std::move(folded), std::move(value));
  }
  return normalized;
}

}  // namespace a11::net

File: kotlin/src/main/kotlin/a11/net/WireStream.kt
/*
 * Copyright 2026 The A11 Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package a11.net

import a11.ByteMap
import a11.Status
import a11.WireMessage
import a11.WritableWireStream

/** Trailer/header carrying a peer's structured non-OK terminal status. */
const val ABORT_STATUS_HEADER = "x-a11-abort-status"

/** Hard ceiling for one reassembled A11 wire message (32 MiB). */
const val MAX_SINGLE_WIRE_MESSAGE_SIZE = 32 * 1024 * 1024

/** Consumes one inbound message; `null` announces the peer's half-close. */
typealias OnWireMessage = suspend (WireMessage?) -> Status

/** Runs once after both directions finish, or after an abort. */
typealias OnWireDone = suspend () -> Status

/**
 * Message-oriented, bidirectional transport shared by sessions and actions.
 *
 * Mirrors the `WireStream` contract in `js/src/wire_stream.ts`: [start]/[accept]
 * drive one endpoint; finish normally with [halfClose] + [drainOutgoingMessages],
 * or [abort] for a failed exchange. Ordered application data travels as sequenced
 * `NodeFragment`s above this layer, so the transport promises no global ordering.
 */
interface WireStream : WritableWireStream {
    override fun send(message: WireMessage): Status
    suspend fun start(onMessage: OnWireMessage? = null, onDone: OnWireDone? = null): Status
    suspend fun accept(onMessage: OnWireMessage? = null, onDone: OnWireDone? = null): Status
    fun halfClose(trailers: ByteMap = ByteMap()): Status
    suspend fun drainOutgoingMessages(): Status
    fun abort(status: Status): Status
    fun getStatus(): Status
    fun getTrailers(): ByteMap?
    fun getId(): String
    suspend fun wait(): Status
}
